### ISSUE TITLE
Difference geoprocessing fails because of GEOS

### ISSUE THREAD

**Issue by arx-it** [qgis/QGIS#39029]
I have a problem with geopackages datas (polygons) : With the "difference" geoprocessing

Here are the data : https://drive.google.com/drive/folders/1jseoFHmzw5r_yQoxSywuPBHZkYctdx63?usp=sharing

I have also reply to this bug report : https://github.com/qgis/QGIS/issues/29400

Here is the log :

Version de QGIS : 3.14.1-Pi
Révision du code : de08d6b
Version de Qt : 5.11.2
Version de GDAL : 3.0.4
Version de GEOS : 3.8.1-CAPI-1.13.3
Version de Proj : Rel. 6.3.2, May 1st, 2020
Traitement de l'algorithme…
Démarrage de l'algorithme 'Différence'…
Paramètres en entrée:
{ 'INPUT' : 'D:/Temp/qgis_crash/grand.gpkg|layername=grand', 'OUTPUT' : 'TEMPORARY_OUTPUT', 'OVERLAY' : 'D:/Temp/qgis_crash/petit.gpkg|layername=petit' }

Erreur de géotraitement GEOS: la différence a échoué.

Échec d'exécution au bout de 0.15 secondes

Chargement des couches de résultat
Algorithme 'Différence' terminé

Thank you in advance.

**Comment by roya0045** [comment]
Are they in the same projection?

**Comment by gioman** [comment]
> Are they in the same projection?

@roya0045 they are.

The operation fails with QGIS native tool, also after running the "fix geometries" tool on the inputs. GRASS and SAGA tools process the data correctly. Maybe this type of problem will be solved when we will, start to use the next gen of GEOS, the one including the improvements about rounding and snapping?

**Comment by gioman** [comment]
Closing as upstream issue?

**Comment by strk** [comment]
Was upstream ticket filed ?

**Comment by gioman** [comment]
> Was upstream ticket filed ?

@strk I don't follow the GEOS tracker.

**Comment by strk** [comment]
The bug is confirmed with GEOS-3.9.0dev:
```
strk=# select ST_Summary(ST_Difference(g.geom, p.geom)) from grand g, petit p;
ERROR:  lwgeom_difference_prec: GEOS Error: TopologyException: found non-noded intersection between LINESTRING (655065 1.79482e+06, 655065 1.79482e+06) and LINESTRING (655065 1.79482e+06, 655065 1.79482e+06) at 655065.0840699079 1794820.6249905184
strk=# select postgis_full_version();
                                                                                                                                                               postgis_full_version                                                                                                                                                               
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 POSTGIS="3.1.0dev 3.1.0alpha2-69-gf50dcaaaa" [EXTENSION] PGSQL="120" GEOS="3.9.0dev-CAPI-1.14.0" PROJ="6.3.1" LIBXML="2.9.10" LIBJSON="0.13.1" LIBPROTOBUF="1.3.3" WAGYU="0.5.0 (Internal)" (core procs from "3.1.0dev 3.1.0alpha2-63-g801577ae4" need upgrade) TOPOLOGY (topology procs from "3.1.0dev 3.1.0alpha2-64-g7e692d0f2" need upgrade)
(1 row)
```

RE following GEOS tracker, what's the policy about upstream issues ? I'd say before closing a ticket here a link to an upstream ticket should be filed ?

**Comment by strk** [comment]
I've filed https://trac.osgeo.org/geos/ticket/1051 for OverlayNG results

**Comment by gioman** [comment]
> 'd say before closing a ticket here a link to an upstream ticket should be filed ?

yes,

thanks for the upstream ticket.

**Comment by arx-it** [comment]
Thank you for your replies.
Ok for GRASS and SAGA tools !

**Comment by strk** [comment]
for the record: this is now fixed in GEOS master (to be released as 3.9.0)

**Comment by arx-it** [comment]
Ok thank you, we will wait QGIS next version
